packages/contracts/contracts/voting/addresslist/AddresslistVoting.sol

```diff
@@ -133,13 +118,7 @@ contract AddresslistVoting is MajorityVotingBase {
             }
         }
 
-        emit ProposalCreated({
-            proposalId: proposalId,
-            creator: _msgSender(),
-            metadata: _proposalMetadata
-        });
-
-        vote(proposalId, _voteOption, _tryEarlyExecution);
+        vote(id, _voteOption, _tryEarlyExecution);
```

Comment:
Why is `vote()` executed unconditionally? Won't it revert if attempted ahead of time?

Reply:
This is the public `vote()` method that has the `_canVote` check inside. All the checks apply here

Reply:
It still doesn't feel logically correct that `createProposal()` calls `vote` unconditionally. 

In addition to this, calling `vote()` as a plain transaction should fail if you cannot vote, whereas it should skip it silently if you cannot vote but call if from `createProposal()`. I highly suggest we rework this hierarchy.

Reply:
@brickpop Could you check the code ? It calls public `vote` function which includes the check whether it can be allowed or not. Just making sure you understand the code.

Reply:
@novaknole but the point is not about the inner logic of `vote()`. It's about the assumptions made when you call it.

One use case assumes "Try to see if you can vote, just in case" and the other assumes "I want to actively vote, tell me if you can't". 

Also, the public `vote()` function is not even defined within the same file.

Reply:
I am against doing calls like "Try to see if you can vote, just in case".

IMO, things that aren't allowed should fail as early as possible.
What if there is an SDK/frontend bug and we accidentally always try to vote on proposal creation?
Nobody would notice and a lot of gas will be wasted.
Or if the proposal creator thought he had multisig approval rights but used the wrong wallet that is not on the address list (and everyone is allowed to create proposals). He/she would be under the assumption that she had approved, but in fact didn't.

Reply:
For the sake of saving gas, I think we should simply skip any other instruction when `_voteOption == VoteOption.None`.

However, in MajorityVoting, we have:
```
function vote(uint256 _proposalId, VoteOption _voteOption, bool _tryEarlyExecution) public {
        address account = _msgSender();

        if (_voteOption != VoteOption.None && !_canVote(_proposalId, account)) {
            revert VoteCastForbidden({proposalId: _proposalId, account: account});
        }
        _vote(_proposalId, _voteOption, account, _tryEarlyExecution);
    }
```

And then, in AddressListVoting, we have:
```
function _vote(
        uint256 _proposalId,
        VoteOption _voteOption,
        address _voter,
        bool _tryEarlyExecution
    ) internal override {
        Proposal storage proposal_ = proposals[_proposalId];

        VoteOption state = proposal_.voters[_voter];

        // Remove the previous vote.
        if (state == VoteOption.Yes) {
            proposal_.tally.yes = proposal_.tally.yes - 1;
        } else if (state == VoteOption.No) {
            proposal_.tally.no = proposal_.tally.no - 1;
        } else if (state == VoteOption.Abstain) {
            proposal_.tally.abstain = proposal_.tally.abstain - 1;
        }

        // Store the updated/new vote for the voter.
        if (_voteOption == VoteOption.Yes) {
            proposal_.tally.yes = proposal_.tally.yes + 1;
        } else if (_voteOption == VoteOption.No) {
            proposal_.tally.no = proposal_.tally.no + 1;
        } else if (_voteOption == VoteOption.Abstain) {
            proposal_.tally.abstain = proposal_.tally.abstain + 1;
        }

        proposal_.voters[_voter] = _voteOption;

        emit VoteCast({
            proposalId: _proposalId,
            voter: _voter,
            voteOption: _voteOption,
            votingPower: 1
        });

        if (_tryEarlyExecution && _canExecute(_proposalId)) {
            _execute(_proposalId);
        }
    }
```

Which means that, even if I'm not voting anything upon creation, we are still going to have to execute `vote()`, then `_vote()`, then store the vote option and finally emit a `Vote` event. Even if the value was `None`.

Wouldn't it be easier to simply check if `_voteOption` is `None` and skip the call to `vote()` if so?

Reply:
Fixed in bde8e977dac0e25adbc36982d49efdd8bfebe318 together with a related bug being connected to voting with `VoteOption.None`.

The bug likely originated from postponing the tests to https://aragonassociation.atlassian.net/browse/APP-1416 while doing structural changes in the voting / multisig classes. The follow-up testing task should improve and do deeper testing also for the other voting-related governance plugins.

Reply:
Fixed in 9a0959ff2a509e3127e92149205249059e5d1210 together with a related bug being connected to voting with `VoteOption.None`.
Replacing a vote back to `None` outside `VoteReplacement` did not cause a revert with `VoteCastForbidden
` (although the tally does, correctly, not change).